### PR TITLE
8336464: C2: Force CastX2P to be a two-address instruction

### PR DESCRIPTION
**Pull request by fg1417**
This patch forces `CastX2P` to be a two-address instruction, so that C2 could allocate the same register for `dst` and `src`. Then we can remove the instruction completely in the assembly.

The motivation comes from some cast operations like `castPP`. The difference for ADLC between `castPP` and `CastX2P` lies in that `CastX2P` always has different types for `dst` and `src`. We can force ADLC to generate an extra `two_adr()` for `CastX2P` like it does automatically for `castPP`, which could tell register allocator that the instruction needs the same register for `dst` and `src`.

However, sometimes, RA and GCM in C2 can't work as we expected.

For example, we have Assembly on the existing code:
```
  ldp    x10, x11, [x17,#136]
  add    x10, x10, x15
  add    x11, x11, x10
  ldr    x12, [x17,#152]
  str    x16, [x10]
  add    x10, x12, x15
  str    x16, [x11]
  str    x16, [x10]
```

After applying the patch independently, the assembly is:
```
  ldr    x10, [x16,#136]  <--- 1
  add    x10, x10, x15
  ldr    x11, [x16,#144]  <--- 2
  mov    x13, x10         <--- 3
  str    x17, [x13]
  ldr    x12, [x16,#152]
  add    x10, x11, x10
  str    x17, [x10]
  add    x10, x12, x15
  str    x17, [x10]
```

C2 generates a totally extra `mov`, see 3, and we even lost the chance to merge load pair, see 1 and 2. That's terrible.

Although this scenario would disappear after combining with https://github.com/openjdk/jdk/pull/20157, I'm still not sure if this patch is worthwhile.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336464](https://bugs.openjdk.org/browse/JDK-8336464): C2: Force CastX2P to be a two-address instruction (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20159/head:pull/20159` \
`$ git checkout pull/20159`

Update a local copy of the PR: \
`$ git checkout pull/20159` \
`$ git pull https://git.openjdk.org/jdk.git pull/20159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20159`

View PR using the GUI difftool: \
`$ git pr show -t 20159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20159.diff">https://git.openjdk.org/jdk/pull/20159.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20159#issuecomment-2233402165)
</details>

### CODE REVIEW

File: src/hotspot/share/adlc/output_h.cpp [hunk 1]
```diff
@@ -1688,6 +1688,15 @@ void ArchDesc::declareClasses(FILE *fp) {
     // Identify which input register matches the input register.
     uint  matching_input = instr->two_address(_globalNames);
 
+#if defined(AARCH64)
+    // Allocate the same register for src and dst, then we can remove
```
[the remark this answers is not in the document]
Comment (fg1417) [reply src/hotspot/share/adlc/output_h.cpp]: Thanks for your review, @theRealAph . Updated it in the new commit.

**Review by vnkozlov — COMMENTED** [review]
I am not sure about this change.  There reason we keep `src `and `dst` in different register for different types is most likely for cases when `src` could be used in other operations. Overwriting `src` register may give you more spills than before.

If there are no other `src` usages RA should handle this I think in shared code.

**Comment by fg1417** [comment]
> I am not sure about this change. There reason we keep `src `and `dst` in different register for different types is most likely for cases when `src` could be used in other operations. Overwriting `src` register may give you more spills than before.
> 
> If there are no other `src` usages RA should handle this I think in shared code.

Thanks for your review @vnkozlov .

In my initial idea, if we keep `src` and `dst` in the same register, when `src` is used in other operations, yes, we need to generate extra spill code like:
```
Spill src to new_src
CastX2P    src src
...  // Other operations use new_src
```
In the final code, we can remove `CastX2P`, it becomes:
```
mov    src new_src
...  // Other operations use new_src
```
If we keep `src` and `dst` in the different registers, we may get:
```
CastX2P    src dst
...  // Other operations use src
```
In the final code, it will be:
```
mov    src  dst
...  // Other operations use src
```
I thought that keeping `src` and `dst` might not generate extra `mov`s because we can remove `CastX2P` itself finally.

But I tried some written cases showed in the PR description, which violated my thoughts in an unexpected way. Then I'm also not sure about it.

Anyway, it's a try and comments are welcome :)

**Comment by vnkozlov** [comment]
One idea I had long time ago [JDK-6768706](https://bugs.openjdk.org/browse/JDK-6768706) is to add mach instructions with complex matching rules which uses CatX2P/CastP2X in common patterns. Then you can avoid moves between registers.

**Comment by dean-long** [comment]
I thought reusing one of the inputs for the destination is the default, and we have to add TEMP to rules to prevent this from happening.  So I don't understand why sometimes the register allocator doesn't reuse the register when there are no other uses.  There is a concept of "chain rule" in ADLC that I don't quite understand, but I suspect that it is related.

**Comment by fg1417** [comment]
Thanks for all your review and comments! I'll come back when I find a better way. Now, I'd like to convert to draft :)

**Comment by theRealAph** [comment]
You may be over-reacting. `ldp` is split by the decoder into two ops in every AArch64 implementation I've seen. Register-register `mov` takes zero cycles on all current AArch64 implementations. (Except perhaps the tiniest, which don't run much Java.) So you are seeing a less efficient encoding, not loss in performance. Granted, it's less efficient in icache and decode bandwidth, but that's not usually a bottleneck.

**Comment by theRealAph** [comment]
On the other hand, the move generated by `CastX2P` also takes zero cycles to execute, so I guess it's a wash.

**Comment by fg1417** [comment]
> You may be over-reacting. `ldp` is split by the decoder into two ops in every AArch64 implementation I've seen. Register-register `mov` takes zero cycles on all current AArch64 implementations. (Except perhaps the tiniest, which don't run much Java.) So you are seeing a less efficient encoding, not loss in performance. Granted, it's less efficient in icache and decode bandwidth, but that's not usually a bottleneck.

Hi @theRealAph, thanks for your comments. That quite makes sense to me and I learnt a lot!

I tried some hand-written assembly benchmarks. As you expected, the performance results of `two-ldr` version and `one-load-pair` version, whether using `32-bit` registers or `64-bit` registers, are very close on the V2 machine. But I checked the [V2 software optimization guide](https://developer.arm.com/documentation/109898/latest/), which says the execution latency and throughput of `w-form ldp` and `ldr` are the same. That confuses me. Does that mean, during the instruction execution, execution latency or throughput are not the main factors in this case?

> On the other hand, the move generated by `CastX2P` also takes zero cycles to execute, so I guess it's a wash.

Yes, there may be no obvious performance uplift after removing these zero latency `mov`s. One thing is the code size generated by C2 on `aarch64` in some workloads is much larger than on `x86`, which is becoming a problem now. Not sure how much benefit it may bring for this problem, if we could improve the scenarios mentioned above by improving the matcher or register allocator.

Thanks.

**Pull request closed without merging**